### PR TITLE
add xmpp-consumer and common configuration

### PR DESCRIPTION
**Pull request by dmfrey**
This will add XMPP support for the Stream Applications.

### CODE REVIEW

**Comment by onobc** [comment]
@dmfrey do you mind adding some context on the description for this proposal? Thanks.

**Comment by dmfrey** [comment]
> @dmfrey do you mind adding some context on the description for this proposal? Thanks.

I updated the description.

**Review by artembilan — CHANGES_REQUESTED** [review]
I believe we cannot merge this until correct Copyright header in every single class.

File: functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java
Comment (artembilan) [inline functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java:28]: I don't think this is OK to accept this real host in our tests...

**Review by artembilan — CHANGES_REQUESTED** [review]

File: functions/common/xmpp-common/src/main/java/org/springframework/cloud/fn/common/xmpp/XmppConnectionFactoryConfiguration.java
Comment (dmfrey) [reply functions/common/xmpp-common/src/main/java/org/springframework/cloud/fn/common/xmpp/XmppConnectionFactoryConfiguration.java:25]: I just added it to be consistent with how `spring-integration-xmpp` was configuring the connection, which defaults to true as well. I'll remove this property.

File: functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java
Comment (dmfrey) [reply functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java:28]: these will be removed. it was just to test against a real server. Gonna try to come up with a docker version that can be loaded with test containers

File: functions/function-dependencies/pom.xml
Comment (dmfrey) [reply functions/function-dependencies/pom.xml:47]: I mentioned them in another issue, and @onobc mentioned they were gonna removed. I'll merge that change in when completed.

File: functions/consumer/xmpp-consumer/src/main/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerProperties.java
Comment (onobc) [inline functions/consumer/xmpp-consumer/src/main/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerProperties.java:37]: We typically do not add javadoc on the setters but instead use the config props description on the field to be the documenter.

**Review by onobc — CHANGES_REQUESTED** [review]
@dmfrey thanks for the contribution. 

I would like to understand the value of adding this though. Is this something that you are using currently (or going to use) or are there others that will use this too? 

I am being cautious about what we add because:
1. every set of apps we add has maintenance costs and 
2. there are only a few dedicated resources on this project and 
3. there are already many apps to take care of 

Don't get me wrong, thank you very much for the proposal. I just want to make sure that the benefit of adding this outweighs the maintenance costs. 

I went ahead and reviewed the code and have requested some minor updates - mostly of the NIT variety.

File: functions/consumer/xmpp-consumer/src/main/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerProperties.java
Comment (dmfrey) [reply functions/consumer/xmpp-consumer/src/main/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerProperties.java:37]: I'll remove the javadocs. I was following what we previously did.

File: functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java
Comment (dmfrey) [reply functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java:52]: It has been removed.

File: functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java
Comment (dmfrey) [reply functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java:65]: it has been marked `private`

File: functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java
Comment (dmfrey) [reply functions/consumer/xmpp-consumer/src/test/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerConfigurationTests.java:71]: it has been removed

File: functions/consumer/xmpp-consumer/src/main/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerProperties.java
Comment (artembilan) [reply functions/consumer/xmpp-consumer/src/main/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerProperties.java:37]: @dmfrey ,

Not sure how others, but I prefer to have changes applied without confirmation via comments.
Otherwise it sounds like a spam in my email box: first commit with the change we have requested then your comments that you have done them 😄

Comment (onobc) [reply functions/consumer/xmpp-consumer/src/main/java/org/springframework/cloud/fn/consumer/xmpp/XmppConsumerProperties.java:37]: I understand wanting to "ack" the requested change via the "this thing is done" but it **can** get spammy on notifications. A couple of less-spammy alternatives that achieve the same purpose are: 
1. when complete is just adding the thumbsup emoji as I don't think that sends a notification (if it does send notification then it would be no better of course and would come off of this list)
2. just resolving the comment - that also sends the signal that it is complete. 
3. summary reply message when "all the requested changes are in".

**Review by onobc — COMMENTED** [review]
@dmfrey everything looks good but I do have some questions around the testcontainers side of things though.

1. As it is, only the consumer function can use this. It would be useful in the other modules as well. What is your plan on that front? 

2. What would this look like (aka tradeoffs) if instead of keeping the openfire conf files locally and docker composing the container at test time we baked an image that has that, pushed it to a registry, then just refefenced that image in a `GenericContainer`? 

Thanks for the updates.

**Comment by dmfrey** [comment]
> @dmfrey everything looks good but I do have some questions around the testcontainers side of things though.
> 
> 1. As it is, only the consumer function can use this. It would be useful in the other modules as well. What is your plan on that front?
> 2. What would this look like (aka tradeoffs) if instead of keeping the openfire conf files locally and docker composing the container at test time we baked an image that has that, pushed it to a registry, then just refefenced that image in a `GenericContainer`?
> 
> Thanks for the updates.

@onobc I'm happy to implement whatever is best. If we build a container out of this, where is the registry where it would be hosted?  Also, I could see moving this configuration to a test-scoped artifact in the `xmpp-common` project that could then be injected into the individual functions, so there is no duplication. Thoughts?

File: functions/consumer/xmpp-consumer/src/test/docker-compose/xmpp/docker-compose.yml
Comment (dmfrey) [reply functions/consumer/xmpp-consumer/src/test/docker-compose/xmpp/docker-compose.yml:6]: This is the standard port it uses for clients to connect. So internally, it will need to be mapped to that port, ultimately.  How do you propose to proceed?

Comment (onobc) [reply functions/consumer/xmpp-consumer/src/test/docker-compose/xmpp/docker-compose.yml:6]: Take a look at the source code for `org.testcontainers.containers.MongoDBContainer` and you will see how they use a mapped port. Another option would be to create an xmpp TC impl like `MongoDBContainer` rather than the docker compose route. 

As for proceeding - this does not have to be done in the scope of this PR as long as we have an issue created to follow up and do that in some reasonable subsequent time period. :) What you have works and we should get this in and iterate on it from there. My 2c.

**Comment by onobc** [comment]
> If we build a container out of this, where is the registry where it would be hosted? Also, I could see moving this configuration to a test-scoped artifact in the xmpp-common project that could then be injected into the individual functions, so there is no duplication. Thoughts?

Yeh, there are downsides of the custom image approach in that regard (which registry etc..). We can leave this in the xmpp-consumer/test for now and as you said move it to xmpp-common when you add the next piece that needs it (as we have done w/ the other testcontainer support we have). You can see xmpp-redis/mongodb for examples of the maven dependnecies etc.. 

#### My recommendation would be:
1. we can merge this as-is. 
2. handle the mapped port concern in subsequent PR
3. handle the "sharing of the thing" (aka move to xmpp-common) in whatever next PR needs the "thing"

If you agree, I will get this merged here shorlty.

**Comment by onobc** [comment]
@dmfrey this is closed via https://github.com/spring-cloud/stream-applications/commit/c39eeee50b03532af4a5631886acd874bf0cff4a

Please follow up on the items in https://github.com/spring-cloud/stream-applications/pull/351#issuecomment-1285747375 when you can. Thanks again.

**Pull request closed without merging**
